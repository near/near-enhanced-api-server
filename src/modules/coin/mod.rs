use actix_web::{HttpResponse, error};
use actix_web_validator::{Error, PathConfig};
use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

#[derive(serde::Serialize)]
pub struct ValidationErrorJsonPayload {
    pub message: String,
    pub fields: Vec<String>,
}

/// Custom error handler
impl From<&validator::ValidationErrors> for ValidationErrorJsonPayload {
    fn from(error: &validator::ValidationErrors) -> Self {
        ValidationErrorJsonPayload {
            message: "Validation error".to_owned(),
            fields: error.field_errors().iter().map(|(field, _)| field.to_string()).collect(),
        }
    }
}

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app
        .app_data(PathConfig::default().error_handler(|err, _| {
            let json_error = match &err {
                Error::Validate(error) => ValidationErrorJsonPayload::from(error),
                _ => ValidationErrorJsonPayload { message: err.to_string(), fields: Vec::new() },
            };
            error::InternalError::from_response(err, HttpResponse::Conflict().json(json_error)).into()
        }))
        .service(
        web::resource("/accounts/{account_id}/coins/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins")
            .route(web::get().to(resources::get_coin_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}")
            .route(web::get().to(resources::get_coin_balances_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    )
    .service(
        web::resource("/accounts/{account_id}/coins/{contract_account_id}/history")
            .route(web::get().to(resources::get_coin_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_contract_metadata)),
    );
}
