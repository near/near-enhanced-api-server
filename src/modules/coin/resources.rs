use paperclip::actix::{
    api_v2_operation,
    web::{self, Json},
};

use super::{data_provider, schemas};
use crate::{db_helpers, errors, modules, types};

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR balance
///
/// This endpoint returns the NEAR balance of the given `account_id`
/// at the given `block_timestamp_nanos`/`block_height`.
pub async fn get_near_balance(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::NearBalanceResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    Ok(Json(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0).await?,
    ))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances
///
/// This endpoint returns all the countable coin balances (including NEAR, FTs)
/// of the given `account_id`, at the given `block_timestamp_nanos`/`block_height`.
/// todo it's not great, we can't keep sorting + pagination when we add MT here
/// NEAR balance always goes first, then the balances are sorted alphabetically by `contract_account_id`.
///
/// You can also provide `limit` (max 100) and `after_contract_account_id`.
/// `after_contract_account_id` is the way to paginate through the values.
///
/// Some pagination notes: leave `after_contract_account_id` empty at your first request;
/// then, take `contract_account_id` of the last provided item in your previous response.
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
    limit_params: web::Query<types::query_params::LimitParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    let mut limit = types::query_params::checked_get_limit(limit_params.limit)?;
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let mut balances: Vec<schemas::Coin> = vec![];
    balances.push(
        data_provider::get_near_balance(&pool, &block, &request.account_id.0)
            .await?
            .into(),
    );
    limit -= 1;

    if limit > 0 {
        let ft_balances = &mut data_provider::get_coin_balances(
            &pool,
            &rpc_client,
            &block,
            &request.account_id.0,
            limit,
        )
        .await?;
        balances.append(ft_balances);
    }

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin balances by contract
///
/// This endpoint returns all the countable coin balances of the given `account_id`,
/// for the given `contract_account_id` and `block_timestamp_nanos`/`block_height`.
/// For FT contracts, the response has only 1 item in the list.
/// For MT contracts, there could be several balances (MT support is still under development).
/// todo change interface for ft/mt
///
/// **Limitations**
/// * We are in the process of supporting Multi Token balances.
pub async fn get_coin_balances_by_contract(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceByContractRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::CoinBalancesResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native balance, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;
    modules::check_account_exists(&rpc_client, &request.account_id.0, block.height).await?;

    let balances = data_provider::get_coin_balances_by_contract(
        &rpc_client,
        &block,
        &request.contract_account_id.0,
        &request.account_id.0,
    )
    .await?;

    Ok(Json(schemas::CoinBalancesResponse {
        balances,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's NEAR history
///
/// This endpoint returns the history of native NEAR operations,
/// for the given `account_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// We are in the process of collecting the data. We support only last 3 months history for now.
pub async fn get_near_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::BalanceRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_near_history(
            &pool_balances.pool,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get user's coin history by contract
///
/// This endpoint returns the history of coin operations (FT, other standards)
/// for the given `account_id`, `contract_id`, going from the latest to the earliest events.
/// For the next page, use `after_event_index` and pass there the index from the last event from the previous page.
///
/// **Limitations**
/// * If the history does not match with the balances from RPC, we do not support the history (see Trusted/Inconsistent Contracts section below).
/// * We are in the process of supporting Multi Token history.
///
/// **Trusted/Inconsistent Contracts**
/// The contracts produce [the balance changing events](https://nomicon.io/Standards/Tokens/FungibleToken/Event).
/// The contract is marked as inconsistent if its events do not match with the balance taken from RPC.
/// The contract is marked as trusted otherwise.
/// More info could be found [here](https://github.com/near/near-indexer-events#my-contract-produces-eventstheres-a-custom-legacy-logic-for-my-contract-but-the-enhanced-api-still-ignores-me-why).
pub async fn get_coin_history(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    pool_balances: web::Data<db_helpers::DBWrapper>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::HistoryRequest>,
    pagination_params: web::Query<types::query_params::PaginationParams>,
) -> crate::Result<Json<schemas::HistoryResponse>> {
    if request.contract_account_id.to_string() == "near" {
        return Err(errors::ErrorKind::InvalidInput(
            "For native history, please use NEAR (uppercase)".to_string(),
        )
        .into());
    }

    let pagination = modules::checked_get_pagination_params(&pagination_params).await?;
    let block = db_helpers::checked_get_block_from_pagination(&pool, &pagination).await?;
    // we don't need to check whether account exists. If not, we can just return the empty history

    Ok(Json(schemas::HistoryResponse {
        history: data_provider::get_coin_history(
            &pool,
            &pool_balances.pool,
            &rpc_client,
            &request.contract_account_id.0,
            &request.account_id.0,
            &block,
            &pagination,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}

#[api_v2_operation(tags(Coins))]
/// Get FT contract metadata
///
/// This endpoint returns the metadata for a given FT contract and `timestamp`/`block_height`.
///
/// **Limitations**
/// * For now, we support only FT contracts that implement the Events NEP standard.
///   We are working on a solution to support other FT contracts, including `wrap.near` and bridged tokens.
pub async fn get_ft_contract_metadata(
    pool: web::Data<sqlx::Pool<sqlx::Postgres>>,
    rpc_client: web::Data<near_jsonrpc_client::JsonRpcClient>,
    _: crate::types::pagoda_api_key::PagodaApiKey,
    request: actix_web_validator::Path<schemas::ContractMetadataRequest>,
    block_params: web::Query<types::query_params::BlockParams>,
) -> crate::Result<Json<schemas::FtContractMetadataResponse>> {
    let block = db_helpers::checked_get_block(&pool, &block_params).await?;

    Ok(Json(schemas::FtContractMetadataResponse {
        metadata: data_provider::get_ft_contract_metadata(
            &rpc_client,
            request.contract_account_id.0.clone(),
            block.height,
        )
        .await?,
        block_timestamp_nanos: types::U64::from(block.timestamp),
        block_height: types::U64::from(block.height),
    }))
}
