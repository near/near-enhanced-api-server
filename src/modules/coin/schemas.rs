use paperclip::actix::Apiv2Schema;
use validator::{Validate, ValidationError};

use crate::types;

// *** Requests ***

// move to coins
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct BalanceByContractRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub account_id: types::AccountId,
    pub contract_account_id: types::AccountId,
}

// duplicate in each folder
#[derive(Validate, Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct ContractMetadataRequest {
    #[validate(custom = "near_primitives::types::AccountId::validate")]
    pub contract_account_id: types::AccountId,
}

// *** Responses ***

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct NearBalanceResponse {
    /// Sum of staked and nonstaked balances
    pub balance: types::U128,
    pub metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response gives the information about all the available balances for the user.
/// The answer gives the list of NEAR, FT balances, could be used for Multi Tokens.
/// For MTs and other standards, balances could have multiple entries for one contract.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinBalancesResponse {
    pub balances: Vec<Coin>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

/// This response provides the coin history (NEAR or by contract).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryResponse {
    pub history: Vec<HistoryItem>,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadataResponse {
    pub metadata: FtContractMetadata,
    pub block_timestamp_nanos: types::U64,
    pub block_height: types::U64,
}

// ---

/// This type describes general coin information.
/// It is used for NEAR, FT balances, could be used for Multi Tokens.
///
/// For MTs and other standards, we could have multiple coins for one contract.
/// For NEAR and FTs, coin_metadata contains general metadata (the only available option, though).
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct Coin {
    /// "nearprotocol" for NEAR, "nep141" for FT
    pub standard: String,
    pub balance: types::U128,
    /// null for NEAR, not null otherwise
    pub contract_account_id: Option<types::AccountId>,
    pub metadata: CoinMetadata,
    // TODO PHASE 1 (idea) I think it would be great to add here the info about last update moment. Timestamp, later also index
    // I'm already doing it at NftCount
}

/// This type describes the history of coin movements for the given user.
/// Coins could be NEAR, FT, it could be also later used for Multi Tokens.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct HistoryItem {
    // TODO PHASE 2 add index here
    // pub index: types::U128,
    // TODO PHASE 1 (idea) do we want to add here tx_hash/receipt_id? We may want to add it at many places
    pub involved_account_id: Option<types::AccountId>,
    pub delta_balance: types::I128,
    pub balance: types::U128,
    pub cause: String,
    pub status: String,
    pub coin_metadata: CoinMetadata,
    pub block_timestamp_nanos: types::U64,
    // TODO PHASE 2 add this when we have all the data in the same DB. Now we can't join with blocks
    // pub block_height: types::U64,
}

/// This type describes general Metadata info, collecting the most important fields from different standards in the one format.
/// `decimals` may contain `0` if it's not applicable (e.g. if it's general MT metadata)
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct CoinMetadata {
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub decimals: u8,
}

/// The type for FT Contract Metadata. Inspired by
/// https://nomicon.io/Standards/Tokens/FungibleToken/Metadata
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, Apiv2Schema)]
pub struct FtContractMetadata {
    pub spec: String,
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub reference: Option<String>,
    pub reference_hash: Option<String>,
    pub decimals: u8,
}

pub fn validate(account_id: &str) -> Result<(), ValidationError> {
    Err(ValidationError::new("something"))
}
