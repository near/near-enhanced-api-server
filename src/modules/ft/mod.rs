use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/FT")
            .route(web::get().to(resources::get_ft_balances)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/FT/{contract_account_id}")
            .route(web::get().to(resources::get_ft_balance_by_contract)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/FT/{contract_account_id}/history")
            .route(web::get().to(resources::get_ft_history)),
    )
    .service(
        web::resource("/nep141/metadata/{contract_account_id}")
            .route(web::get().to(resources::get_ft_metadata)),
    );
}
