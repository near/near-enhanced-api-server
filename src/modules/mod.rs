use crate::{db_helpers, errors, types};

pub(crate) mod coin;
pub(crate) mod nft;

pub(crate) async fn check_account_exists(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    account_id: &near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<()> {
    let request = near_jsonrpc_client::methods::query::RpcQueryRequest {
        block_reference: near_primitives::types::BlockId::Height(block_height).into(),
        request: near_primitives::views::QueryRequest::ViewAccount {
            account_id: account_id.clone(),
        },
    };
    for _ in 0..5 {
        match rpc_client.call(&request).await {
            Err(near_jsonrpc_client::errors::JsonRpcError::ServerError(
                near_jsonrpc_client::errors::JsonRpcServerError::HandlerError(
                    near_jsonrpc_client::methods::query::RpcQueryError::UnknownAccount { .. },
                ),
            )) => {
                return Err(errors::ErrorKind::InvalidInput(format!(
                    "account_id {} does not exist at block_height {}",
                    account_id, block_height
                ))
                .into())
            }
            Err(err) => {
                tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence via JSON RPC failed with: {:?}. Re-trying immediatelly", err);
                continue;
            }
            Ok(response) => {
                if let near_jsonrpc_primitives::types::query::QueryResponseKind::ViewAccount(_) =
                    response.kind
                {
                    return Ok(());
                } else {
                    tracing::warn!(target: crate::LOGGER_MSG, "Checking account existence returned invalid response: {:?}. Re-trying immediatelly", response);
                    continue;
                }
            }
        }
    }
    Err(errors::ErrorKind::InternalError(format!(
        "could not check if account_id {} exists after several attemps",
        account_id
    ))
    .into())
}

pub(crate) async fn check_and_get_history_pagination_params(
    pool: &sqlx::Pool<sqlx::Postgres>,
    pagination_params: types::query_params::HistoryPaginationParams,
) -> crate::Result<types::query_params::HistoryPagination> {
    types::query_params::check_limit(pagination_params.limit)?;
    let pagination = types::query_params::Pagination::from(pagination_params);
    // if pagination_params.after_block_height.is_some() && pagination_params.after_timestamp_nanos.is_some() {
    //     return Err(errors::ErrorKind::InvalidInput(
    //         "Both block_height and block_timestamp_nanos found. Please provide only one of values"
    //             .to_string(),
    //     )
    //         .into());
    // }
    // TODO PHASE 2 take the block from pagination_params
    let block = db_helpers::get_last_block(pool).await?;
    Ok(types::query_params::HistoryPagination {
        block_height: block.height,
        block_timestamp: block.timestamp,
        limit: pagination.limit,
    })
}

#[cfg(test)]
mod tests {
    use crate::db_helpers;

    pub(crate) async fn init_db() -> sqlx::Pool<sqlx::Postgres> {
        dotenv::dotenv().ok();
        let db_url = &std::env::var("DATABASE_URL").expect("failed to get database url");

        sqlx::PgPool::connect(db_url)
            .await
            .expect("failed to connect to the database")
    }

    pub(crate) fn init_rpc() -> near_jsonrpc_client::JsonRpcClient {
        dotenv::dotenv().ok();
        let rpc_url = &std::env::var("RPC_URL").expect("failed to get RPC url");
        let connector = near_jsonrpc_client::JsonRpcClient::new_client();
        connector.connect(rpc_url)
    }

    pub(crate) fn get_block() -> db_helpers::Block {
        db_helpers::Block {
            timestamp: 1655571176644255779,
            height: 68000000,
        }
    }
}
