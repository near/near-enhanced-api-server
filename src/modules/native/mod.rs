use paperclip::actix::web;

mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/accounts/{account_id}/balances/NEAR")
            .route(web::get().to(resources::get_near_balance)),
    )
    .service(
        web::resource("/accounts/{account_id}/balances/NEAR/history")
            .route(web::get().to(resources::get_near_history)),
    );
}
