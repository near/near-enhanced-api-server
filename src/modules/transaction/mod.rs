use paperclip::actix::web;

// mod data_provider;
mod resources;
mod schemas;

pub(crate) fn register_services(app: &mut web::ServiceConfig) {
    app.service(
        web::resource("/transaction/{transaction_hash}")
            .route(web::get().to(resources::get_transaction_by_tx_hash)),
    )
    .service(
        web::resource("/transaction/receipt/{receipt_id}")
            .route(web::get().to(resources::get_transaction_by_receipt)),
    )
    .service(
        web::resource("/transactions/{account_id}")
            .route(web::get().to(resources::get_transactions_by_account)),
    )
    .service(
        web::resource("/transactions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_transactions_by_account_on_contract)),
    )
    .service(
        web::resource("/receipts/{transaction_hash}")
            .route(web::get().to(resources::get_receipts_by_tx_hash)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}")
            .route(web::get().to(resources::get_action_receipts_by_account)),
    )
    .service(
        web::resource("/receipts/actions/{account_id}/{contract_id}")
            .route(web::get().to(resources::get_action_receipts_by_account_on_contract)),
    );
}
