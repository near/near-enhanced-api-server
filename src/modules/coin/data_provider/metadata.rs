use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};

use crate::modules::coin;
use crate::{rpc_helpers, types};

pub(crate) async fn get_ft_contract_metadata(
    rpc_client: &near_jsonrpc_client::JsonRpcClient,
    contract_id: near_primitives::types::AccountId,
    block_height: u64,
) -> crate::Result<coin::schemas::FtContractMetadata> {
    let request = rpc_helpers::get_function_call_request(
        block_height,
        contract_id.clone(),
        "ft_metadata",
        serde_json::json!({}),
    );
    let response =
        rpc_helpers::wrapped_call(rpc_client, request, block_height, &contract_id).await?;

    let metadata = serde_json::from_slice::<FtMetadata>(&response.result)?;
    Ok(coin::schemas::FtContractMetadata {
        spec: metadata.spec,
        name: metadata.name,
        symbol: metadata.symbol,
        icon: metadata.icon,
        decimals: metadata.decimals,
        reference: metadata.reference,
        reference_hash: types::vector::base64_to_string(&metadata.reference_hash)?,
    })
}

pub(crate) fn get_default_metadata() -> coin::schemas::CoinMetadata {
    coin::schemas::CoinMetadata {
        name: "The contract did not provide the metadata".to_string(),
        symbol: "The contract did not provide the symbol".to_string(),
        icon: None,
        decimals: 0,
    }
}

pub(crate) fn get_near_metadata() -> coin::schemas::CoinMetadata {
    coin::schemas::CoinMetadata {
        name: "NEAR blockchain native token".to_string(),
        symbol: "NEAR".to_string(),
        // TODO PHASE 2 re-check the icon. It's the best I can find
        icon: Some("https://raw.githubusercontent.com/near/near-wallet/7ef3c824404282b76b36da2dff4f3e593e7f928d/packages/frontend/src/images/near.svg".to_string()),
        decimals: 24,
    }
}

// Taken from https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/fungible_token/metadata.rs
#[derive(BorshDeserialize, BorshSerialize, Clone, Deserialize, Serialize, Debug)]
pub struct FtMetadata {
    pub spec: String,
    pub name: String,
    pub symbol: String,
    pub icon: Option<String>,
    pub reference: Option<String>,
    pub reference_hash: Option<types::vector::Base64VecU8>,
    pub decimals: u8,
}

impl From<coin::schemas::FtContractMetadata> for coin::schemas::CoinMetadata {
    fn from(metadata: coin::schemas::FtContractMetadata) -> Self {
        coin::schemas::CoinMetadata {
            name: metadata.name,
            symbol: metadata.symbol,
            icon: metadata.icon,
            decimals: metadata.decimals,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::tests::*;
    use std::str::FromStr;

    // todo add test on bridged contracts after we decide something on https://github.com/near/near-enhanced-api-server/issues/43
    #[tokio::test]
    async fn test_ft_contract_metadata() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("usn").unwrap();

        let metadata = get_ft_contract_metadata(&rpc_client, contract, block.height).await;
        insta::assert_debug_snapshot!(metadata);
    }

    #[tokio::test]
    async fn test_ft_contract_metadata_no_contract_deployed() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("olga.near").unwrap();

        let metadata = get_ft_contract_metadata(&rpc_client, contract, block.height).await;
        insta::assert_debug_snapshot!(metadata);
    }

    #[tokio::test]
    async fn test_ft_contract_metadata_other_contract_deployed() {
        let rpc_client = init_rpc();
        let block = get_block();
        let contract = near_primitives::types::AccountId::from_str("comic.paras.near").unwrap();

        let metadata = get_ft_contract_metadata(&rpc_client, contract, block.height).await;
        insta::assert_debug_snapshot!(metadata);
    }
}
