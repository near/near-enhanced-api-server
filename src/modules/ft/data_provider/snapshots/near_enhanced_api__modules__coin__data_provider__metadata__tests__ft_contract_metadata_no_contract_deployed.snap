---
source: src/modules/coin/data_provider/metadata.rs
expression: metadata
---
Err(
    Error {
        code: 400,
        message: "Invalid Input: The account `olga.near` does not implement any suitable contract at block 68000000",
        retriable: false,
    },
)
